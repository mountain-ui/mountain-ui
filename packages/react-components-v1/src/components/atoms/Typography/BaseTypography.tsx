import { MntProps, createMntComponent } from '../../../internals/create_mnt_component';

export interface BaseTypographyProps extends MntProps {
  /* Text should be uncopyable */
  uncopyable?: boolean;
  /* Text should be bold */
  bold?: boolean;
  /* Text should be wrapped in ellipsis */
  truncate?: boolean;
  /* Text should be clamped to specified line count */
  lineClamp?: number;
  /* Text should be underlined */
  underline?: boolean;
  /* Text have gradient color */
  gradient?: TextGradientOptions;
}

export interface TextGradientOptions {
  from: string;
  to: string
  direction?: 't' | 'tr' | 'r' | 'br' | 'b' | 'bl' | 'l' | 'tl'
}

const bold = "font-bold"
const gradient = (opts?: TextGradientOptions) => opts ? `text-transparent bg-clip-text bg-gradient-to-${opts.direction} from-${opts.from} to-${opts.to}` : ''
const lineClamp = (rowsCount?: number) => rowsCount > 0 ? `line-clamp-${rowsCount}` : ''
const truncate = 'truncate'
const uncopyable = 'select-none'
const underline = 'underline underline-offset-2';

const BaseTypography = createMntComponent<BaseTypographyProps>('span')({
  classFromProps: {
    bold,
    gradient,
    lineClamp,
    truncate,
    uncopyable,
    underline,
  }
})

BaseTypography.displayName = 'BaseTypography';

export default BaseTypography;