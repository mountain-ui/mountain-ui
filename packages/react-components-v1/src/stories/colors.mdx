import { Meta, ColorPalette, ColorItem } from '@storybook/blocks';

<Meta title='Design System/Colors' />

# Colors

For the color palette, we inherit the [Tailwind color palette](https://tailwindcss.com/docs/customizing-colors), adding some custom colors based on out needs.

<ColorPalette>
  <ColorItem title='Light' colors={['hsl(0, 0%, 98%)']} />
  <ColorItem title='Dark' colors={['hsl(210, 14%, 9%)']} />
</ColorPalette>
