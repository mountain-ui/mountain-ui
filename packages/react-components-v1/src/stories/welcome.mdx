import { Meta } from '@storybook/blocks';

<Meta title='Design System/Welcome' />

<h1 className='mb-2 text-h1 text-green-400'>Mountain UI 🏔</h1>
<p mb={2}>
  This tool helps us explore, text and document all the components from the Mountain UI design
  system.
</p>
<p mb={2}>
  It is divided into more stories, you can find them on the left navigation sidebar. Start clicking
  a story to play around with the library components.
</p>
