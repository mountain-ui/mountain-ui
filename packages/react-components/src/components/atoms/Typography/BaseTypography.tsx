import styled, { css } from 'styled-components';
import {
  compose,
  system,
  textStyle,
  TextStyleProps,
  typography,
  TypographyProps
} from 'styled-system';

import { BaseElement, BaseElementProps } from '../../BaseElement';

const textStyles = compose(typography, textStyle);

export type BaseTypographyProps = BaseElementProps &
  TypographyProps &
  TextStyleProps & {
    /* Text should be uncopyable */
    uncopyable?: boolean;
    /* Text should be bold */
    strong?: boolean;
    /* Text should be wrapped in ellipsis */
    wrapped?: boolean;
    /* Text should be clamped to specified line count */
    lineClamp?: number;
    /* Text should be underlined */
    underline?: boolean;
    /* Text have gradient color */
    gradient?: string;
  };

const wrapped = css`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

const lineClamp = css`
  display: -webkit-box;
  -webkit-box-orient: vertical;
  overflow: hidden;
`;

const strong = css`
  font-weight: ${p => p.theme.fontWeights.bold};
`;

const underline = css`
  text-decoration: underline;
`;

const uncopyable = css`
  user-select: none;
`;

const gradientColor = css`
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
`;

/**
 * The `BaseTypography` component works as the first brick to compose more specific text components
 */
const BaseTypography = styled(BaseElement)<BaseTypographyProps>`
  ${textStyles}
  ${system({ userSelect: true })}
  ${p => p.gradient && `background: linear-gradient(${p.gradient}); ${gradientColor} `}
  ${p => p.lineClamp && `${lineClamp}-webkit-line-clamp: ${p.lineClamp};`}
  ${p => p.strong && strong}
  ${p => p.uncopyable && uncopyable}
  ${p => p.underline && underline}
  ${p => p.wrapped && wrapped}
`;
BaseTypography.displayName = 'BaseTypography';

export default BaseTypography;
