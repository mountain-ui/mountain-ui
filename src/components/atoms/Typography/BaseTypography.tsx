import styled, { css } from 'styled-components';
import { compose, system, textStyle, TextStyleProps, typography, TypographyProps } from 'styled-system'

import { BaseElement, BaseElementProps } from '../../../BaseElement';

const textStyles = compose(typography, textStyle)

export type BaseTypographyProps = BaseElementProps & TypographyProps & TextStyleProps & {
  /* Text should be unselectable */
  unselectable?: boolean;
  /* Text should be bold */
  strong?: boolean;
  /* Text should be wrapped in ellipsis */
  wrapped?: boolean;
  /* Text should be clamped to specified line count */
  lineClamp?: number;
  /* Text should be underlined */
  underline?: boolean;
}

const wrapped = css`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

const lineClamp = css`
  display: -webkit-box;
  -webkit-box-orient: vertical;
  overflow-y: hidden;
`;

const strong = css`
  font-weight: ${p => p.theme.fontWeights.bold};
`;

const underline = css`
  text-decoration: underline;
`;

const unselectable = css`
  user-select: none;
`;

/**
 * TODO: add component description headline
 */
const BaseTypography = styled(BaseElement) <BaseTypographyProps>`
  ${textStyles}
  ${system({ userSelect: true })}
  ${p => p.wrapped && wrapped}
  ${p => p.strong && strong}
  ${p => p.lineClamp && `${lineClamp}-webkit-line-clamp: ${p.lineClamp};`}
  ${p => p.underline && underline}
  ${p => p.unselectable && unselectable}
  ${p => `
    ${p.wrapped && wrapped}
    ${p.strong && strong}
    ${p.lineClamp && `${lineClamp}-webkit-line-clamp: ${p.lineClamp}`}
    ${p.underline && underline}
    ${p.unselectable && unselectable}
  `}
`
BaseTypography.displayName = 'BaseTypography';

export default BaseTypography