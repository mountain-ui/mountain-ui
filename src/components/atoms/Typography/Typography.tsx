import styled, { css } from 'styled-components';
import { compose, textStyle, TextStyleProps, typography, TypographyProps as TypoProps, variant } from 'styled-system'

import { BaseElement, BaseElementProps } from '../../../BaseElement';

type TextProps = TypoProps & TextStyleProps

const textStyles = compose(typography, textStyle)

export type TypographyProps = BaseElementProps & TextProps & {
  /**
   * Text should be wrapped in ellipsis
   */
  wrapped?: boolean;
  /**
   * Text should be clamped to specified line count
   */
  lineClamp?: number;
}

export type HeadingProps = TypographyProps & {
  /**
   * Heading variant
   */
  variant: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
}

const wrapped = css`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

const lineClamp = css`
  display: -webkit-box;
  -webkit-box-orient: vertical;
  overflow-y: hidden;
`;

/**
 * TODO: add component description headline
 */
const BaseTypography = styled(BaseElement) <TypographyProps>`
  ${textStyles}
  ${p => p.wrapped && wrapped}
  ${p => p.lineClamp && `${lineClamp}-webkit-line-clamp: ${p.lineClamp};`}
`
export default BaseTypography

export const Heading = styled(BaseTypography).attrs((p: HeadingProps) => ({ as: p.as || p.variant }))<HeadingProps>(
  variant({
    scale: 'variants.typography.heading',
    variants: {
      h1: {
        color: 'text.primary',
        fontSize: 'h1',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h2: {
        color: 'text.primary',
        fontSize: 'h2',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h3: {
        color: 'text.primary',
        fontSize: 'h3',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h4: {
        color: 'text.primary',
        fontSize: 'h4',
        lineHeight: 'normal',
        fontWeight: 'bold'
      },
      h5: {
        color: 'text.primary',
        fontSize: 'h5',
        lineHeight: 'normal',
        fontWeight: 'bold'
      },
      h6: {
        color: 'text.primary',
        fontSize: 'h6',
        lineHeight: 'normal',
        fontWeight: 'bold'
      }
    }
  })
)

Heading.defaultProps = {
  variant: 'h1'
}

export const Text = styled(BaseTypography).attrs((p: HeadingProps) => ({ as: p.variant }))<HeadingProps>(
  variant({
    scale: 'variants.typography.heading',
    variants: {
      h1: {
        color: 'text.primary',
        fontSize: 'h1',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h2: {
        color: 'text.primary',
        fontSize: 'h2',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h3: {
        color: 'text.primary',
        fontSize: 'h3',
        lineHeight: 'short',
        fontWeight: 'bold'
      },
      h4: {
        color: 'text.primary',
        fontSize: 'h4',
        lineHeight: 'normal',
        fontWeight: 'bold'
      },
      h5: {
        color: 'text.primary',
        fontSize: 'h5',
        lineHeight: 'normal',
        fontWeight: 'bold'
      },
      h6: {
        color: 'text.primary',
        fontSize: 'h6',
        lineHeight: 'normal',
        fontWeight: 'bold'
      }
    }
  })
)

Heading.defaultProps = {
  variant: 'h1'
}

BaseTypography.displayName = 'BaseTypography';