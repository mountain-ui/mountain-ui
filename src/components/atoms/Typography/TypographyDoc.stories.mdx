import { Meta, Typeset, ArgsTable } from '@storybook/addon-docs/blocks';
import theme from '../../../defaultTheme';

<Meta title='Atoms/Typography/Sizes' />

## Font Sizes

<Typeset
  fontSizes={theme.fontSizes}
  fontWeight={theme.fontWeights.normal}
  fontFamily={theme.fonts.normal}
/>
