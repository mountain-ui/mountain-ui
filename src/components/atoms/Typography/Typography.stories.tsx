import React from 'react'
import { Meta } from '@storybook/react/types-6-0';

import { Container } from '../Layout';

import BaseTypography, { Heading } from './Typography';


export default {
  title: 'Atoms/Typography',
  component: BaseTypography,
  argTypes: {

  }
} as Meta;

export const HeadingStory = () => <Container>
  <Heading variant="h1">Mountain UI library</Heading>
  <Heading variant="h2">Mountain UI library</Heading>
  <Heading variant="h3">Mountain UI library</Heading>
  <Heading variant="h4">Mountain UI library</Heading>
  <Heading variant="h5">Mountain UI library</Heading>
  <Heading variant="h6">Mountain UI library</Heading>
  <BaseTypography fontSize="body">Mountain UI library</BaseTypography>
  <BaseTypography fontSize="secondaryBody">Mountain UI library</BaseTypography>
  <BaseTypography fontSize="sub">Mountain UI library</BaseTypography>
  <BaseTypography fontSize="caption">Mountain UI library</BaseTypography>
</Container>

HeadingStory.storyName = 'Heading'

HeadingStory.parameters = {
  jest: ['Typography.test.js'],
};
