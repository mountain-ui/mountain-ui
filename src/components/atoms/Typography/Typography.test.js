import React from 'react';
import initStoryshots from '@storybook/addon-storyshots';
import { render } from '@testing-library/react';

import { HeadingStory } from './Typography.stories';

initStoryshots();

describe('Typography', () => {
  describe('Heading component', () => {
    it('should render correctly on mount', () => {
      const { container } = render(<HeadingStory {...HeadingStory.args} />);
      expect(container).toBeInTheDocument();
    });
  });
});
