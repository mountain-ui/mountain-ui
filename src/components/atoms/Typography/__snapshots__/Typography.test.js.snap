// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Layout Container 1`] = `
<div
  className="sc-bdfBwQ sc-gsTCUz bFhuNW kNNlrj"
  height="100vh"
>
  This is a responsive container
</div>
`;

exports[`Storyshots Atoms/Layout Grid 1`] = `
<div
  className="sc-bdfBwQ sc-dlfnbm bFhuNW dYmuJF"
  height="100vh"
>
  This is a responsive Grid
</div>
`;

exports[`Storyshots Atoms/Typography Heading 1`] = `
<div
  className="sc-bdfBwQ sc-gsTCUz  kNNlrj"
>
  <h1
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC kEjxqG"
  >
    Mountain UI library
  </h1>
  <h2
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC jAzxnd"
  >
    Mountain UI library
  </h2>
  <h3
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC jmGPhr"
  >
    Mountain UI library
  </h3>
  <h4
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC gzQKTa"
  >
    Mountain UI library
  </h4>
  <h5
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC jMLdHy"
  >
    Mountain UI library
  </h5>
  <h6
    className="sc-bdfBwQ sc-hKgILt sc-eCssSg  cTxziC dROxyE"
  >
    Mountain UI library
  </h6>
  <div
    className="sc-bdfBwQ sc-hKgILt  eIleIv"
    fontSize="body"
  >
    Mountain UI library
  </div>
  <div
    className="sc-bdfBwQ sc-hKgILt  dIKksG"
    fontSize="secondaryBody"
  >
    Mountain UI library
  </div>
  <div
    className="sc-bdfBwQ sc-hKgILt  laatu"
    fontSize="sub"
  >
    Mountain UI library
  </div>
  <div
    className="sc-bdfBwQ sc-hKgILt  XsOPt"
    fontSize="caption"
  >
    Mountain UI library
  </div>
</div>
`;
